plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java

```diff
@@ -168,8 +169,14 @@ public static Type createTypeFromCoercer(TypeManager typeManager, HiveType fromH
         if (fromType == REAL && toType instanceof DecimalType toDecimalType) {
             return Optional.of(createRealToDecimalCoercer(toDecimalType));
         }
-        if (fromType instanceof TimestampType && toType instanceof VarcharType varcharType) {
-            return Optional.of(new LongTimestampToVarcharCoercer(TIMESTAMP_NANOS, varcharType));
+        if (fromType instanceof TimestampType) {
+            if (toType instanceof VarcharType varcharType) {
+                return Optional.of(new LongTimestampToVarcharCoercer(TIMESTAMP_NANOS, varcharType));
+            }
+            if (toType instanceof DateType toDateType) {
+                return Optional.of(new LongTimestampToDateCoercer(TIMESTAMP_NANOS, toDateType));
```

Comment:
This seems incorrect.  The from type is a timestamp type, but we are passing nanos to `LongTimestampToDateCoercer`.  That code will call `getObject` on the type to get `LongTimestamp`.  The probem is that method will only work if the Block type is Fixed12Block, but if the input data is millis, the block will be a LongArrayBlock.

Reply:
Actually when Hive tries to tries to coerces a timestamp column to a varchar or date - we forcefully convert the read type to `TIMESTAMP_NANOS` - Ad the underling block would be of `Fixed128Block` always.

Hive doesn't have the concept of parameterized timestamp (atleast in Hive 3) so all the column would be read as `TIMESTAMP(9)`so we handle for coercion accordingly.

Reply:
For reference to the above answer:

https://github.com/trinodb/trino/blob/36895bb93aa054536148cb47d8220c616399c2d0/plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java#L84-L89

Reply:
Tested extensively manually with combinations of :+1: 

- `2021-01-01 00:00:00`
- `2021-01-02 00:00:00.123`
- `2021-01-03 00:00:00.123456`
- `2021-01-04 00:00:00.123456789`

and each time we read `timestamp(9)` which ends up being stored in Fixed12Block

Reply:
Why "we forcefully convert the read type to TIMESTAMP_NANOS" if we are going to just convert back to millis or micros (depending on the configuration of the catalog)? Seems pretty wasteful.

Add a comment explaining this behavior and maybe add a pointer to the code that changes the type to `timestamp(9)`

Reply:
> we forcefully convert the read type to TIMESTAMP_NANOS

We are reading the column with `TIMESTAMP_NANOS` only if some coercion is applied on top of it - During all the other case we read it as per the configuration. We could implement another coercer for short timestamp - but now we have support reading timestamp column with NANOS precision are we going to deprecate `HiveTimestampPrecision` in the future ?

Reply:
> Why "we forcefully convert the read type to TIMESTAMP_NANOS" if we are going to just convert back to millis ?

Probably because we don't want to suffer from side effects of rounding `2023-12-31 23:59:59.999999999` to `2024-01-01 00:00:00.000`

https://github.com/trinodb/trino/blob/5d167d4570e99bb052f9130efc07854c047c91d6/lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java#L399-L406